opencda/core/ml_libs/reinforcement_learning/envs/simple_carla_env.py

```diff
@@ -0,0 +1,726 @@
+import os
+import time
+import carla
+import torch
+import numpy as np
+from typing import Any, Dict, Optional, Tuple
+import math
+import gym
+from gym import spaces
+from collections import defaultdict
+
+from opencda.core.ml_libs.reinforcement_learning.simulators import CarlaSimulator
+from opencda.core.ml_libs.reinforcement_learning.utils.env_utils.stuck_detector import StuckDetector
+from opencda.core.ml_libs.reinforcement_learning.utils.simulator_utils.carla_utils import lane_mid_distance
+
+from ding.utils.default_helper import deep_merge_dicts
+from ding.envs.env.base_env import BaseEnvTimestep, BaseEnvInfo
+from ding.envs.common.env_element import EnvElementInfo
+from ding.torch_utils.data_helper import to_ndarray
+
+from abc import ABC, abstractmethod
+import copy
+from easydict import EasyDict
+from gym import utils
+
+# class SimpleCarlaEnv(BaseDriveEnv):
+class SimpleCarlaEnv(gym.Env, utils.EzPickle):
+    """
+    A simple deployment of Carla Environment with single hero vehicle. It use ``CarlaSimulator`` to interact with
+    Carla server and gets running status. The observation is obtained from simulator's state, information and
+    sensor data, along with reward which can be calculated and retrived.
+
+    When created, it will initialize environment with config and Carla TCP host & port. This method will NOT create
+    simulator instance. It only creates some data structures to store information when running env.
+
+    :Arguments:
+        - cfg (Dict): Env config dict.
+        - host (str, optional): Carla server IP host. Defaults to 'localhost'.
+        - port (int, optional): Carla server IP port. Defaults to 9000.
+        - tm_port (Optional[int], optional): Carla Traffic Manager port. Defaults to None.
+
+    :Interfaces: reset, step, close, is_success, is_failure, render, seed
+
+    :Properties:
+        - hero_player (carla.Actor): Hero vehicle in simulator.
+    """
+
+    metadata = {'render.modes': ['rgb_array']}
```

Comment:
I don't prefer the style that puts code before __init__, maybe DI drive did such thing, but let's avoid this